### DISCUSSION TITLE
Import history data

### DISCUSSION THREAD

**Discussion by markcocker** [BottlecapDave/HomeAssistant-OctopusEnergy#899] [General]
Is it possible to import historical electric and gas data from Octopus into HA long term statistics, either with this integration or manually?

I came across the HA integration https://github.com/klausj1/homeassistant-statistics/ that can import a .CSV file but it requires data to be hourly, and the Octopus website only gives an option to download your data in half-hourly slots.

**Comment by BottlecapDave** [comment]
Hello. All previous consumption sensors push data into long term statistics in order to support the energy dashboard at a granualar level. The names of these sensors can be found in the [energy dashboard setup](https://bottlecapdave.github.io/HomeAssistant-OctopusEnergy/setup/energy_dashboard/#previous-day-consumption).

If you are wishing to import data before you set up the integration, you can use the [refresh service](https://bottlecapdave.github.io/HomeAssistant-OctopusEnergy/services/#octopus_energyrefresh_previous_consumption_data), however this will take a while and depending on how far back you go, you might hit some API limits.

**Comment by markcocker** [comment]
Hi. 

For electricity, when I call the [refresh service](https://bottlecapdave.github.io/HomeAssistant-OctopusEnergy/services/#octopus_energyrefresh_previous_consumption_data) with a start date of 2024-01-01 it returns successfully and I get a notification but after an hour I don't see data for January, and don't see an error in the log.

For gas the instructions says to use sensor: 

`sensor.octopus_energy_gas_{{METER_SERIAL_NUMBER}}_{{MPRN_NUMBER}}_previous_accumulative_consumption`

which appears successful, but the log says `Referenced entities sensor.octopus_energy_gas_e6s16545252261_8818917908_current_accumulative_consumption are missing or not currently available`

I only see one sensor with trailing `_kwh` and another with `_m3` :

`sensor.octopus_energy_gas_e6s16545252261_8818917908_previous_accumulative_consumption_kwh`
`sensor.octopus_energy_gas_e6s16545252261_8818917908_previous_accumulative_consumption_m3`

I use either of those I get a `Failed to call service octopus_energy.refresh_previous_consumption_data. Unknown error`

and the log shows...

```
Logger: homeassistant.components.websocket_api.http.connection
Source: helpers/service.py:1038
integration: Home Assistant WebSocket API ([documentation](https://www.home-assistant.io/integrations/websocket_api), [issues](https://github.com/home-assistant/core/issues?q=is%3Aissue+is%3Aopen+label%3A%22integration%3A+websocket_api%22))
First occurred: 15:24:01 (8 occurrences)
Last logged: 16:05:49

[547638860896] Error handling message: Unknown error (unknown_error) Mark from 192.168.0.1 (Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36)
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/decorators.py", line 27, in _handle_async_response
    await func(hass, connection, msg)
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 793, in handle_execute_script
    script_result = await script_obj.async_run(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 1773, in async_run
    return await asyncio.shield(create_eager_task(run.async_run()))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 464, in async_run
    await self._async_step(log_exceptions=False)
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 528, in _async_step
    self._handle_exception(
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 558, in _handle_exception
    raise exception
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 526, in _async_step
    await getattr(self, handler)()
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 763, in _async_call_service_step
    response_data = await self._async_run_long_action(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 726, in _async_run_long_action
    return await long_task
           ^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2741, in async_call
    response_data = await coro
                    ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2784, in _execute_service
    return await target(service_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 977, in entity_service_call
    single_response = await _handle_entity_call(
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 1038, in _handle_entity_call
    partial(getattr(entity, func), **data),  # type: ignore[arg-type]
            ^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'OctopusEnergyCurrentAccumulativeGasConsumptionKwh' object has no attribute 'async_refresh_previous_consumption_data'
```

**Comment by BottlecapDave** [comment]
> For electricity, when I call the [refresh service](https://bottlecapdave.github.io/HomeAssistant-OctopusEnergy/services/#octopus_energyrefresh_previous_consumption_data) with a start date of 2024-01-01 it returns successfully and I get a notification but after an hour I don't see data for January, and don't see an error in the log.

Do you get two notifications? You should get one for it started and one to state its finished. This will download consumption and rate information a day at a time (in my testing, anything more than this resulted in errors). I've been noticing today the API timing out a fair amount, which might be related. What happens if you don't go back quite as far (e.g. the start of the month)?

> For gas the instructions says to use sensor:
sensor.octopus_energy_gas_{{METER_SERIAL_NUMBER}}_{{MPRN_NUMBER}}_previous_accumulative_consumption
which appears successful, but the log says Referenced entities sensor.octopus_energy_gas_e6s16545252261_8818917908_current_accumulative_consumption are missing or not currently available

The logs you supplied suggest you're using the current accumulative sensor not the previous accumulative sensor.

You should be using `sensor.octopus_energy_gas_e6s16545252261_8818917908_previous_accumulative_consumption_m3` (i've updated the docs for future releases).

**Reply by markcocker** [reply]
Thank you for updating the docs.

I can now call the refresh service for both my gas and electricity sensors, I get notifications that they start then finish the import, and no errors in the `System > Logs > Home Assistant Core`. However, the last image shows no data shows in the history graph. I have tried refreshing the page and logging out/in. I have multiple Octopus accounts setup and checked there is no data for any of those for this period.

* sensor.octopus_energy_electricity_21l4726127_2000021792976_previous_accumulative_consumption
* sensor.octopus_energy_gas_e6s16545252261_8818917908_previous_accumulative_consumption_m3

<img width="475" alt="image" src="https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/15351000/a9cb81ea-29ba-4703-bbff-e29c6cd97ea7">

<img width="1549" alt="image" src="https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/15351000/d924d34d-a5e5-49ae-afb1-717f463cf95e">

**Reply by BottlecapDave** [reply]
So this doesn't import data into the sensors themselves (this is due to HA entities not liking historic data being added retroactively). As mentioned the data is imported into the special statistic sensors mentioned in the energy dashboard docs (as this is the purpose of the service)
